### ISSUE TITLE
[core] IME Support

### ISSUE THREAD

**Issue by mia-bentzen** [raysan5/raylib#1945]
In order to be able to type text in japanese (and i assume chinese too, though i'm not familiar with it) you pretty much have to use to OS's IME (Input MEthod) to get input working right.

To give a small example of how it works: when you type が you don't actually type が directly, rather you type ga and it then automatically replaces that with が. It also comes up with suggestions based on a dictionary, and pressing space changes kana (basically letters) into kanji (basically complicated letters) so for instance you could have おんな, press space, and then it would change it into 女.

In order for the OS to know where and when to draw the IME suggestions box, the application has to tell it. The way SDL2 accomplishes this is with three methods: SDL_StartTextInput, SDL_StopTextInput, and SDL_SetTextInputRect.

The reason this would be nice to have is that without it japanese users simply can't type japanese text in any raylib application, and i'm not aware of any alternative or workaround.

**Comment by orcmid** [comment]
That's very interesting.  I was thinking that we need a way to do that and there are also considerations with respect to accessibility.  This is also valuable for additional languages/keyboards.

The TLDR: This software is extensive and maybe larger than raylib :).  Figuring out a mapping and then being able to deal with SDL2 updates is going to be challenging.  I recommend that this issue be moved to Discussion since adapting SDL for inter-working with raylib is going to require extensive effort and substantial analysis and testing.  

=== Observations

The [SDL2 Tutorial](https://www.matsson.com/prog/sdl2-mingw-w64-tutorial.php) is tied to mingw-w64 and it involves libraries as well as a DLL that applications are expected to bundle.  It is useful that the code downloads are signed.  There is no public development server (GitHub or elsewhere).

It may take some work to adapt that to the raylib way of doing things, whether as an external or extra.  I assume that raylib should be moved to handle Unicode (probably UTF-8) in these cases.

I gave the source-code download a quick look.  The code is under a BSD-flavor license.  Considering that this is about a Windows-centric facility (I thought) the plethora of Unix-isms is startling (*.sh, *.m4, etc.) in areas involving use of Visual C..   Apparently there are comparable integrations for other platforms.   There are Visual Studio solutions and projects as part of the code.  Some may be dated (one is for VS2012).  The dates on files in the Zip appear to be from late July although it is not clear what that means.  There is no `.git` and change-control history.

**Comment by mia-bentzen** [comment]
First off, I might be misunderstanding what you're saying, so absolutely correct me if i'm wrong!

You seem to be suggesting that we use SDL2 to solve this problem, while that might be possible my intention was more just to show what the API might look like (i'm not really experienced enough with this kind of thing to be much help with the low level details anyway).

Another thing is that IMEs are not a windows thing, they're on mac and linux as well. Otherwise you wouldn't be able to type in japanese on either of those.

As for whether or not it should be moved to discussion i don't really know how any of that works (this is the first issue i've ever opened). So if i need to do something then i might need instructions.

**Comment by orcmid** [comment]
Thank you @mia-bentzen, I understand your concern more now.  が　I managed to use IME right here, after using Windows 10 to install Japanese as an additional language.  I did not know how to do the rest of your example, of course, although bringing up special soft keyboards might help..  

I am uncertain that my browser has anything to do with this with regard to enabling IME.  I am ignorant of how the interaction is handled.  

Two things:

 1. The integration model for enabling IME (if that is needed), or allowing it to be shown, needs to be understood better.  I hope you can help us understand that.
 2. To change this issue to a discussion, there should be an option on the right sidebar here to do that.  I don't see it, probably because I am not the author of the issue.  Ideally, it will take the entire thread along with it.

**Comment by mia-bentzen** [comment]
So i suppose there's two parts to understanding how the IME works. There's the way the user interacts with it, which i explained a bit, and the more technical details (i can only really give educated guesses here)

For completeness i'll try to give as complete of an explanation of both as i can.

So first the user has to select a text box of some kind, which then prompts the IME to become enabled. You can then activate it by switching to hiragana or katakana. Once that happens if you start typing it will start giving auto complete suggestions (this is definitely handled by the OS, as it's very consistent in it's behavior and you can customize the dictionary in the language settings somewhere).

The user can type the romaji (english letters representing the japanese character) of any kana. So for instance my earlier example of ga being turned into が. At this point the kana that was typed is in a sort of limbo state where it gets displayed by the application, but it hasn't been sent as input as the OS is in full control (or at least mostly) of what is happening to it.

Once you hit enter you commit those characters and it gets sent to the application as text that has been inputted, and the application is free to handle it however it wants. 

this process can be viewed very intuitively in sublime text because for some reaseon it actually shows the 'limbo' text in a separate box.
![raylib_ime_gif](https://user-images.githubusercontent.com/11667218/130509182-ec55bb25-1e17-4912-a130-9b45441ffa8c.gif)

Another thing that you can do is press space (only during the limbo state) to 'collapse' the kana into kanji. This is very important as kanji is extremely common in japanese.

Based on all of this, i think that during the limbo state the application only gets told what the current 'limbo' text is, and then once you commit all that gets sent as some kind of text input event (i'm not familiar with how this is handled in the low level OS APIs, so it might either be multiple events for each character or one event with the entire string). And the application needs to tell the OS when a textbox (could be anything really, but the point is that you type text there and it has a location) is selected that the IME needs to start and where it needs to start. (To clarify, when i say start here i don't mean opening the suggestions, that is done by the user typing. Rather i mean that we tell the OS that IME input is now possible).

Apologies for the wall of text, i figured it was better to be too detailed than not detailed enough however. If anything was unclear just tell me and i'll try and reexplain that part better!

(Also, i don't seem to have the option to change the issue to a discussion either, i don't know if i'm looking at the wrong place or doing something wrong or something else)

**Comment by orcmid** [comment]
Aside: When you view this issue, the right sidebar should have something like this at the end of the material there:
![ZoomResult](https://user-images.githubusercontent.com/3372580/130529137-39bedc69-0853-4d94-9f95-6a0152635f35.png)

As the originator of the issue, I think it should appear for you.  I am uncertain whether it brings all of your nice comments along.  I would hope so.

**Comment by mia-bentzen** [comment]
For whatever reason none of those buttons are there for me.

**Comment by raysan5** [comment]
@mia-bentzen Thanks for the detailed report, this issue could require some investigation.

@orcmid I think only the owner of the repo can convert an Issue to a Discussion. For me it's ok to keep it as an issue.

**Comment by orcmid** [comment]
I have been sleuthing in SDL.  Here's what I have so far:

Aside: To conduct this analysis I had to install [grepWin](https://tools.stefankueng.com/grepWin.html).  Any good `grep` can be used to search an entire project.  It is a quick way to narrow in.  Sometimes there are too many hits though.

The functions `SDL_StartTextInput`, `SDL_StopTextInput`, and `SDL_SetTextInputRec` and others have their signatures specified in the file `SDL_keyboard.h` around line 231 and down through to the end.

The functions `SDL_StartTextInput`, `SDL_StopTextInput`, `SDL_SetTextInputRec`, and soft-keyboard displaying are implemented in `SDL_video.c` at lines 4034-4106.

These are wrappers that rely on a global variable, `_this`, defined at `SDL_video.c` line 128.

```
	static SDL_VideoDevice *_this = NULL;
```

The `struct SDL_VideoDevice` pointed to is a structure of pointers to functions defined in `SDL_sysvideo.h`.  An `SDL_VideoDevice`  is a very large _vTable_ of sorts, a kind of vector into functions implemented elsewhere.  The functions pointed to in an initialized `SDL_VideoDevice struct` must all take a `_THIS` parameter, where

```
		 #define _THIS  SDL_VideoDevice *_this	
 ```
 
 is the parameter type used in the prototypes of various functions.  This is demonstrated in the code for `SDL_StartTextInput` which uses the 
 structure member `void (*StartTextInput)(_THIS_)` to access a platform-specific implementation:
 
 ```
 	_this->StartTextInput(_this);
 ```
 
 The point is that the functions pointed to in the `SDL_VideoDevice` are initialized with platform-specific solutions, depending on what platform (e.g., Linux, Windows, MacOS) SDL has been compiled for.  The use of providing the  _vTable_ as a parameter makes all of the specific `SDL_VideoDevice struct` available to each of the functions linked from it and the table could even be altered by them.  This is a kind of roll-your own virtual class (C++) expressed in C language.  I will have more to say about that.
 
 There are 7 relevant text-input and possible-screen-keyboard functions that have pointers stashed in an `SDL_VideoDevice`.  The slots have the following names and prototypes for the functions pointed to.
 
 ```
  /* Text input */

 void (*StartTextInput) (_THIS);

 void (*StopTextInput) (_THIS);

 void** (*SetTextInputRect) (_THIS, SDL_Rect *rect);


 /* Screen keyboard */

 SDL_bool (*HasScreenKeyboardSupport) (_THIS);

 void (*ShowScreenKeyboard) (_THIS, SDL_Window *window);

 void (\*HideScreenKeyboard) (_THIS, SDL_Window *window);

 SDL_bool (*IsScreenKeyboardShown) (_THIS, SDL_Window *window);
 ```
 
There are checks on whether a pointer is present, as in 

```
if (! _this->HasScreenKeyboard ) ... ;
```

And if it is available, the function is used via the `_this` pointer as in

```
if (! _this->HasScreenKeyboard )
		 return SDL_false; 
	else return this->HasScreenKeyboard(_this);
```

We have seen how the keyboard functions are implemented by short procedures that call for platform-specific implementations via the global `_this` pointer to an initialized `struct SDL_VideoDevice`.

Now we need to understand where such initializations are made and what are the functions pointed to after initialization.

------

Note 1: This is a bit like hand-rolled COM interfacing.  It is weirder than that in the extreme use of indirection, even though `_this` is likely held in registers by compiler optimization.  There is reference counting in SDL2, and in-process COM would provide a general solution for that as well.  One would not encapsulate all of this behind a single COM interface such as the mammoth `SDL_VideoDevice` `struct`.  I speculate that better functional encapsulation is achievable.  Likewise if this was done with C++ abstract classes.

Note 2: raylib also relies on a global data structure, `CORE`, which means essentially that an application can have only one (main) window.  At least `CORE` is directly reached rather than via a global pointer.  Current reliance on a DLL (or other shared library) for SDL2 appears to be a problem if the DLL ends up being shared between two running programs at the same time.  It can work.  It seems unnecessarily fragile.  There is a different philosophy with `raylib`, where the application and `raylib` modules are compiled together and all loaded into the created executable (e.g., an `.exe`)

**Comment by jmorel33** [comment]
Fantastic dive and information https://github.com/orcmid not everyone would be willing to dig into.  Thanks for getting your hands dirty on this.

**Comment by orcmid** [comment]
Continuing ...

One place where initialization of a `SDL_VideoDevice struct` occurs is in the function 

```
	static SDL_VideoDevice *
		WIN_CreateDevice(int devindex)
```

start ing at line 97 of `SDL_windowsvideo.c`.  Here a `struct VideoDevice` is allocated and then populated.   The function will set `WIN_StartTextInput`, `WIN_StopTextInput`, and `Win_SetTextInputRect` in the corresponding structure slots  `StartTextInput`, `StopTextInput`, and `SetTextInputRect`.  There are no settings made for soft keyboards.  The `devindex` parameters is not used.

The `Win_CreatDevice` function is not called directly - it is private to the `SDL_windowsvideo.c`.  However, its address is included in the data declaration.

```
	VideoBootStrap WINDOWS_bootstrap = {

		"windows", "SDL Windows video driver", WIN_CreateDevice };
 ```
 
 The `typedef` for type `struct VideoBootStrap` is defined in `sysvideo.h` and the `WINDOWS_bootstrap` is made `external` there.
 
 In `SDL_video.c`  the `&WINDOWS_bootstrap` address, if used, will be made an entry in an array of bootstrap pointers.
 
 This pattern applies for each of the video systems supported in SDL2.  We will follow the Windows case out of consistency and familiarity with how IME's are handled.
 
 Assume that `Win_CreateDevice` is called.  Then the interesting functions are `WIN_StartTextInput`, `WIN_StopTextInput`, and `WIN_SetTextInputRect`.   These are defined in `windowskeyboard.c`  They are relatively long, including the other functions called, and I have  no idea where this fits in how input is processed in raylib.  The length seems a function of being compiled to include IME support (the default).
 
 I think the hard part remains.  It is beyond my capability and capacity.

**Comment by UradaSources** [comment]
> * [IME input glfw/glfw#41](https://github.com/glfw/glfw/issues/41)
> * [Add IME support for each platform glfw/glfw#2130](https://github.com/glfw/glfw/pull/2130)

Thanks a lot! This turned out to be supported in glfw. I'll look into it later. But raylib itself doesn't seem to have an api like StartInput? Will this be added in the future?

**Comment by raysan5** [comment]
@jdeokkim It seems those changes are not merged yet but I reopen this issue to keep track of it.

**Comment by UradaSources** [comment]
> @jdeokkim @UradaSources It seems those changes are not merged yet but I reopen this issue to keep track of it.

I took a quick look at the glfw api and did some tests, just simply register a callback function with glfwSetCharCallback to get the output of ime! This callback function is called once for each complete character returned by ime, and the codepoint is the corresponding character unicode code.

**Comment by daipom** [comment]
> > * [IME input glfw/glfw#41](https://github.com/glfw/glfw/issues/41)
> > * [Add IME support for each platform glfw/glfw#2130](https://github.com/glfw/glfw/pull/2130)
> 
> Thanks a lot! This turned out to be supported in glfw. I'll look into it later. But raylib itself doesn't seem to have an api like StartInput? Will this be added in the future?

Hi, I am one of those working on this GLFW PR.

* [Add IME support for each platform glfw/glfw#2130](https://github.com/glfw/glfw/pull/2130)

We are already working on the corresponding work for raylib.

* raylib: https://github.com/clear-code/raylib/tree/better-ime-support
* sample application: https://github.com/clear-code/RaylibIMEInputSampleApp

We are planning to make a PR for raylib once GLFW is merged.

Or if it seems better to set up a PR first, We will make one soon.

**Comment by jdeokkim** [comment]
@daipom The sample application looks very nice, thank you for your hard work!

**Comment by raysan5** [comment]
@daipom Wow! The example is very impressive! Congratulations!

Thank you very much for working on this issue, this is an amazing improvement.

Feel free to send a PR when ready and let's see the complexity/dependencies it adds to raylib and the better way to implement it... In any case, we can put it under a compilation flag for users requiring it.

**Comment by daipom** [comment]
@jdeokkim  @raysan5 Thanks! 
It may take a while to be merged in GLFW.
If it is not a problem, we would like to send a PR in draft status and request a review in advance.

**Comment by raysan5** [comment]
@jmorel33 Feel free to open a PR in draft, also useful for further discussion about this feature.

**Comment by daipom** [comment]
@raysan5 Thanks! I will open the PR in draft soon!

**Comment by daipom** [comment]
I have created the drat PR!

**Comment by jdeokkim** [comment]
@daipom 👏👏👏👏👏👏👏

**Comment by raysan5** [comment]
@daipom Any update on the state of this feature? We are waiting for GLFW merging, right?

**Comment by daipom** [comment]
@raysan5 Yes! We are waiting for GLFW merging!

**Comment by jdeokkim** [comment]
I'd really love to see it getting merged as soon as possible! 🥰

**Comment by daipom** [comment]
Still waiting for GLFW.

**Comment by raysan5** [comment]
@daipom Yeah, no worries... This new GLFW 3.4 release is taking quite long. We should wait...

**Comment by raysan5** [comment]
I'm afraid I'm closing this issue and the related PR. It's been 1.5 years since it was openend and many things changed and a lot of review is required. It's sad because it was great but we should move on. It could be checked back in the future if it applies.
